entity_parser.go

```diff
@@ -241,242 +109,402 @@ func TableFromInstance(object DomainObject) (*Table, error) {
+	}
+
+	var primaryKey *PrimaryKey
+	if val, ok := tagsMap["primaryKey"]; ok {
+		primaryKey, err = parsePrimaryKey(val)
+		if err != nil {
+			return "", NoTTL(), EtlOff, nil, errors.Wrapf(err, "struct %s has invalid primaryKey tag %q", structName, dosaAnnotation)
 		}
 	}
+	return name, ttl, etlState, primaryKey, nil
 }
 
 // parseIndexTag functions parses DOSA index tag
 func parseIndexTag(indexName, dosaAnnotation string) (string, *PrimaryKey, []string, error) {
+	tagsMap, err := getTags(dosaAnnotation)
+	if err != nil {
+		return "", nil, nil, errors.Wrapf(err, "dosa.Index on object %s with an invalid annotation %q", indexName, dosaAnnotation)
+	}
+	if len(tagsMap) == 0 && dosaAnnotation != "" {
+		return "", nil, nil, errors.Wrapf(err, "dosa.Index on object %s with an invalid annotation %q", indexName, dosaAnnotation)
+	}
+
 	// index name struct must be exported in the entity,
 	// otherwise it will be ignored when upserting the schema.
 	if len(indexName) != 0 && unicode.IsLower([]rune(indexName)[0]) {
 		expected := []rune(indexName)
 		expected[0] = unicode.ToUpper(expected[0])
-		return "", nil, nil, fmt.Errorf("index name (%s) must be exported, "+
-			"try (%s) instead", indexName, string(expected))
-	}
-	tag := dosaAnnotation
-
-	// find the primaryKey
-	matchs := indexKeyPattern0.FindStringSubmatch(tag)
-	if len(matchs) != 4 {
-		return "", nil, nil, fmt.Errorf("dosa.Index %s with an invalid dosa index tag %q", indexName, tag)
-	}
-	pkString := matchs[1]
-	key, err := parsePrimaryKey(indexName, pkString)
-	if err != nil {
-		return "", nil, nil, errors.Wrapf(err, "struct %s has an invalid index key %q", indexName, pkString)
+		return "", nil, nil, fmt.Errorf("index name (%s) must be exported, try (%s) instead", indexName, string(expected))
 	}
-	toRemove := strings.TrimSuffix(matchs[0], matchs[2])
-	toRemove = strings.TrimSuffix(matchs[0], matchs[3])
-	tag = strings.Replace(tag, toRemove, "", 1)
 
-	//find the name
-	fullNameTag, name, err := parseNameTag(tag, indexName)
+	name, err := parseNameTag(indexName, tagsMap)
 	if err != nil {
-		return "", nil, nil, errors.Wrapf(err, "invalid name tag: %s", tag)
+		return "", nil, nil, errors.Wrapf(err, "index %s has invalid name tag %q", indexName, dosaAnnotation)
 	}
-	tag = strings.Replace(tag, fullNameTag, "", 1)
 
-	// find the columns
-	fullColumnsTag, indexColumns, err := parseColumnsTag(tag)
+	columns, err := parseColumnsTag(tagsMap)
 	if err != nil {
-		return "", nil, nil, errors.Wrapf(err, "invalid columns tag: %s", tag)
+		return "", nil, nil, errors.Wrapf(err, "index %s has invalid columns tag %q", indexName, dosaAnnotation)
 	}
-	tag = strings.Replace(tag, fullColumnsTag, "", 1)
 
-	tag = strings.TrimSpace(tag)
-	if tag != "" {
-		return "", nil, nil, fmt.Errorf("index field %s with an invalid dosa index tag: %s", indexName, tag)
+	var key *PrimaryKey
+	if val, ok := tagsMap["key"]; ok {
+		key, err = parsePrimaryKey(val)
```

Comment:
Given that we use the same function for paring primary keys and index keys, how about calling it "parseKey"